contracts/BillingConnector.sol

```diff
@@ -0,0 +1,155 @@
+ * through the GRT token bridge.
+ */
+contract BillingConnector is IBillingConnector, Governed {
+    // -- State --
+    // The contract for interacting with The Graph Token
+    IERC20 private immutable graphToken;
+    // The L1 Token Gateway through which tokens are sent to L2
+    ITokenGateway public l1TokenGateway;
+    // The Billing contract in L2 to which we send tokens
+    address public l2Billing;
+
+    /**
+     * @dev Address of the L1 token gateway was updated
+     */
+    event L1TokenGatewayUpdated(address l1TokenGateway);
+    /**
+     * @dev Address of the L2 Billing contract was updated
+     */
+    event L2BillingUpdated(address l2Billing);
+    /**
+     * @dev Tokens sent to the Billing contract on L2
+     */
+    event TokensSentToL2(address _from, address _to, uint256 _amount);
+    /**
+     * @dev Tokens rescued by the governor
+     */
+    event TokensRescued(address indexed to, address indexed token, uint256 amount);
+
+    /**
+     * @dev Constructor function
+     * @param _l1TokenGateway   L1GraphTokenGateway address
+     * @param _l2Billing Address of the Billing contract on L2
+     * @param _token     Graph Token address
+     * @param _governor  Governor address
+     */
+    constructor(
+        address _l1TokenGateway,
+        address _l2Billing,
+        IERC20 _token,
+        address _governor
+    ) Governed(_governor) {
+        _setL1TokenGateway(_l1TokenGateway);
+        _setL2Billing(_l2Billing);
+        graphToken = _token;
+    }
+
+    /**
+     * @dev Sets the L1 token gateway address
+     * @param _l1TokenGateway New address for the L1 token gateway
+     */
+    function setL1TokenGateway(address _l1TokenGateway) external override onlyGovernor {
+        _setL1TokenGateway(_l1TokenGateway);
+    }
+
+    /**
+     * @dev Sets the L2 Billing address
+     * @param _l2Billing New address for the L2 Billing contract
+     */
+    function setL2Billing(address _l2Billing) external override onlyGovernor {
+        _setL2Billing(_l2Billing);
+    }
+
+    /**
+     * @dev Add tokens into the billing contract on L2, for any user
+     * Ensure graphToken.approve() is called for the BillingConnector contract first
+     * @param _to  Address that tokens are being added to
+     * @param _amount  Amount of tokens to add
+     * @param _maxGas Max gas for the L2 retryable ticket execution
+     * @param _gasPriceBid Gas price for the L2 retryable ticket execution
+     * @param _maxSubmissionCost Max submission price for the L2 retryable ticket
+     */
+    function addToL2(
+        address _to,
+        uint256 _amount,
+        uint256 _maxGas,
+        uint256 _gasPriceBid,
+        uint256 _maxSubmissionCost
+    ) external payable override {
+        require(_amount != 0, "Must add more than 0");
+        require(_to != address(0), "destination != 0");
+        require(graphToken.transferFrom(msg.sender, address(this), _amount), "Add transfer failed");
```

Comment:
This might be a good opportunity to use the Permit feature from the GRT. We could add a `addToL2WithSig` so we can get the user to do everything with no token approval transaction, sign message, then send final transaction.